docker/drone/src/main/java/org/arquillian/cube/docker/graphene/location/DockerCubeCustomizableURLResourceProvider.java

```diff
@@ -0,0 +1,335 @@
+        }
+    }
+
+    private String resolveContext(String url) {
+        int indexOfSlash = url.indexOf('/');
+
+        if (indexOfSlash > 0) {
+            return url.substring(indexOfSlash + 1);
+        }
+
+        return "";
+    }
+
+    private int resolvePort(String url) {
+        int port = extractPort(url);
+        return resolveBindPort(port);
+    }
+
+    private int resolveBindPort(int port) {
+        final CubeDockerConfiguration cubeDockerConfiguration = cubeDockerConfigurationInstance.get();
+        final DockerCompositions dockerContainersContent = cubeDockerConfiguration.getDockerContainersContent();
+        final Set<Map.Entry<String, CubeContainer>> containers = dockerContainersContent.getContainers().entrySet();
+
+        if (port == NO_PORT) {
+
+            //if no port, we check if there is only one cube with one bind port and if not, return default one
+
+            int bindPort = -1;
+            for (Map.Entry<String, CubeContainer> cubeContainerEntry : containers) {
+
+                // need to skip vnc and selenium container
+                final SeleniumContainers seleniumContainers = seleniumContainersInstance.get();
+                if (isSeleniumContainers(cubeContainerEntry, seleniumContainers)) {
+                    continue;
+                }
+
+                final CubeContainer cubeContainer = cubeContainerEntry.getValue();
+                if (hasMoreThanOneBindPort(cubeContainer)) {
+                    throw new IllegalArgumentException("No port was specified and a container has more than one bind port.");
+                }
+
+                if (hasOnlyOneBindPort(cubeContainer)) {
+                    if (noPreviousBindPortFound(bindPort)) {
+                        bindPort = cubeContainer.getPortBindings()
+                                .iterator().next()
+                                .getBound();
+                    } else {
+                        throw new IllegalArgumentException("No port was specified and in all containers there are more than one bind port.");
+                    }
+                }
+            }
+
+            if (noPreviousBindPortFound(bindPort)) {
+                throw new IllegalArgumentException("No port was provided and there isn't any bind port.");
+            }
+
+            return bindPort;
+
+        } else {
+            // user specified an exposed port
+            int bindPort = -1;
+            for (Map.Entry<String, CubeContainer> cubeContainerEntry : containers) {
+
+                // need to skip vnc and selenium container
+                final SeleniumContainers seleniumContainers = seleniumContainersInstance.get();
+                if (isSeleniumContainers(cubeContainerEntry, seleniumContainers)) {
+                    continue;
+                }
+
+                final CubeContainer cubeContainer = cubeContainerEntry.getValue();
+                final Collection<PortBinding> portBindings = cubeContainer.getPortBindings();
```

Comment:
I thought the same, the problem is that this class and his aggregator classes contains a lot of `Collections`, and not all of them are used always. So I thought the same of you, but then we are going to instantiate a lot of `ArrayList` that are not going to be used, but they need to be destroyed. So I decided to do that kind of approach instead of creating all the lists. 

Another option could be only create an `ArrayList` for these two arguments, but I don't like this since there are some attributes initialized and anothers not.